### ISSUE TITLE
nuScenes_3Ddetction infer

### ISSUE THREAD

**Issue by lyxbyr** [dlunion/tensorRTIntegrate#9]
你好，这个库已经实现了onnx转tensorrt加速？

**Comment by dlunion** [comment]
是的，可以参照一些文档

**Comment by dlunion** [comment]
是的

**Comment by lyxbyr** [comment]
你好，在Ubuntu下编译，makefile相应的路径改好了， 但make run -j32报错,
已装probuf 3.11C++版本  终端输出很多undefined reference to `google::protobuf::RepeatedPtrField<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::~RepeatedPtrField()'
最后
collect2: error: ld returned 1 exit status
Makefile:62: recipe for target 'workspace/trtrun' failed
make: *** [workspace/trtrun] Error 1

**Comment by lyxbyr** [comment]
您好！thanks you amazing works
我想实现3d检测的推理，nuScenes_3Ddetction.fp32.trtmodel已经生成了，nuScenes_3Ddetction.onnx的输出只比你的nuScenes_3Dtracking.onnx少一个tracking的输出，输入是一张RGB图，我看了您的代码，里面的实现是基于2d检测和跟踪的,3d的检测或跟踪并没用实现，我现在取出了trt forward()后的7个输出了，
eg:auto outHM = boundingbox3DDetect_->tensor("hm")等等;后面代码的三个for循环后不知道怎么改，因为先前您转的2d 检测和跟踪（onnx）最后输出都有个hm_pool, 所以里面的有句	if (*ohmptr == *ohmpoolptr && *ohmpoolptr > threshold)，我这个3d 检测onnx模型没有，我把这个if语句删了，打印这些std::cout << dx << " " << dy << " " << dw << " " << dh与图片的目标数不一致，请问下您怎么改，说的比较乱，如果直接用这些代码推理的话
auto engine = TRTInfer::loadEngine("nuScenes_3Ddetction.fp32.trtmodel");
float mean[3] = {0.485, 0.456, 0.406};
float std[3] = {0.229, 0.224, 0.225};
Mat image = imread("img.jpg");
engine->input()->setNormMat(0, image, mean, std);
engine->forward();
engine->output(0)->print()后续的取值请问下怎么做？谢谢

**Comment by hopef** [comment]
hm是centerNet的heatmap的输出，这个没问题
hm_pool，是因为centerNet一般后处理是需要经过sigmoid和maxpool，如果把这些后处理放到tensorRT时候去做比较麻烦并且性能没有保证，因此我在导出onnx的时候，会修改centerNet的代码，例如：
原始代码：
```
def forward(self, x):
    x = self.base(x)
   ...
   hm = self.hm_head(x)
   wh = self.wh_head(x)
   reg = self.reg_head(x)
   return hm, wh, reg
```

修改后代码：
```
def forward(self, x):
    x = self.base(x)
   ...
   hm = self.hm_head(x)
   wh = self.wh_head(x)
   reg = self.reg_head(x)

   hm = hm.sigmoid()
   hm_pool = F.max_pool2d(hm, 3, padding=1, stride=1)
   return hm, wh, reg, hm_pool
```
也因此会得到hm_pool，所以建议对应修改来简化代码同时得到更高的处理速度

engine->output(0)指第0个输出Tensor，而engine->tensor("hm")指的是hm这个名字的tensor
auto out = engine->output(0);
out->cpu<float>();   # 这句话能够得到这个tensor的指针，并且把数据从GPU复制到CPU
out->width()、height()、channel()、num()，能得到各个维度，具体可以参照头文件

**Comment by lyxbyr** [comment]
恩，谢谢您的解答，我看了下您的预处理代码和Python代码有些不同，3d detection的时候有影响吗？

**Comment by lyxbyr** [comment]
非常感谢～

**Comment by lyxbyr** [comment]
您好，对于sigmoid函数后对于max_pool大小的设定，我看您先前的2D检测或跟踪的代码里面
hm_pool = F.max_pool2d(hm, kernel_size=5, stride=1, padding=2)
先前您回复我的对于3D检测输出的hm,hm_pool = F.max_pool2d(hm, 3, padding=1, stride=1),这个max_pool大小不同,会有影响吗？ 还有depth后处理也需要sigmod函数加偏移，depth.sigmoid()后，也设置这样大小的max_pool ? F.max_pool2d(hm, 3, padding=1, stride=1)

**Comment by lyxbyr** [comment]
您好，因为计算 item['loc'], item['rot_y']需要得到ct值，根据网络的输出是否有amodel_offset分支，ct的计算方式不同，对于3d检测模型有amodel_offest分支，执行如下代码内部的if语句，这条 ct_output = dets['bboxes'][i][j].reshape(2, 2).mean(axis=0) ，dets['bboxes'][i][j]是通过heap topk后经过一系列的转化得到的，如果要得到这个值，是必须要用C++写一个后处理操作(topk,一系列转化)才能得到dets['bboxes'][i][j]吧，为了得到这个值是不是代价有点大(会很慢)？如果不要amodel_offset这分支，走else语句，直接通过box得到ct，又怕最后3d detection会有偏差

`

 if 'rot' in dets and 'dep' in dets and 'dim' in dets \
        and len(dets['dep'][i]) > j:

        if 'amodel_offset' in dets and len(dets['amodel_offset'][i]) > j:
          ct_output = dets['bboxes'][i][j].reshape(2, 2).mean(axis=0)
          amodel_ct_output = ct_output + dets['amodel_offset'][i][j]
          ct = transform_preds_with_trans(
            amodel_ct_output.reshape(1, 2), trans).reshape(2).tolist()
          #print(ct)
        else:
            bbox = item['bbox']
            print(bbox)
            ct = [(bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2]
        item['ct'] = ct
        item['loc'], item['rot_y'] = ddd2locrot(
          ct, item['alpha'], item['dim'], item['dep'], calibs[i]) `




如果要得到dets['bboxes'][i][j]感觉必须重写这些函数？
` heat = _nms(heat)
  scores, inds, clses, ys0, xs0 = _topk(heat, K=K)`

**Comment by dlunion** [comment]
获取中心点，是需要做nms（其实是max_pool2d），然后topk的
而nms里面的max_pool2d可以使用tensorRT内置实现的效率比较好
c++只需要实现，max_pool后的heatmap和原始heatmap是否一致，是则保留，这种留下局部极值
具体的cuda实现，参考[这里ct_detect_backend.cu](https://github.com/dlunion/tensorRTIntegrate/blob/master/src/infer/ct_detect_backend.cu)，通过指定maxobjs后，那么cuda核执行下来只剩下maxobjs个，比如1000，然后对其做排序取topk，时间是比较少的（可以忽略？）

因此只要合理得当，不用担心太慢，不会慢的

**Comment by lyxbyr** [comment]
感谢您的耐心解答，谢谢，我看了ct_detect_backend.cu的实现，里面是实现了您所说的topK，我还是不知道怎么在这里怎么实现求得3D的ct = dets['bboxes'][i][j]，Python代码实现这个是通过返回topk的索引再进行一些操作得到的，但ct_detect_backend.cu里面好像不知道topK的索引？下面Python代码bboxes是求得ct的结果(用到了topk的索引)
` scores, inds, clses, ys0, xs0 = _topk(heat, K=K)
 cts = torch.cat([xs0.unsqueeze(2), ys0.unsqueeze(2)], dim=2)
 xs = xs0.view(batch, K, 1) + reg[:, :, 0:1]
 ys = ys0.view(batch, K, 1) + reg[:, :, 1:2]
 bboxes = torch.cat([xs - wh[..., 0:1] / 2, 
                        ys - wh[..., 1:2] / 2,
                        xs + wh[..., 0:1] / 2, 
                        ys + wh[..., 1:2] / 2], dim=2)

`

**Comment by dlunion** [comment]
你理解错了，ct_detect_backend.cu只是完成了_nms的作用，并没有做topk。他那不是topk
[这里是](https://github.com/dlunion/tensorRTIntegrate/blob/2a8cad8eef1234e352fa370e3d8ff8af2dff8f1b/src/examples/center_net_coco2x_dcn.cpp#L48)完整的cpp代码实现。实现了_nms，但没有topk。但是复原了框的信息，根据这种类似的你能够恢复你想要的3d信息的。因为他是空间对应的，不需要你所谓的什么topk索引

**Comment by lyxbyr** [comment]
嗯嗯，谢谢，我直接用 ct = [(bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2]代替了，可视化后感觉也没多大差别，相当于网络就没用到'amodel_offset'这个输出了， 在1080ti上测试了基本上整个过程要跑61ms左右，请问下您是在什么服务器测试的？

**Comment by hopef** [comment]
我使用的是2080Ti，1080Ti的速度确实挺慢的，主要在于dcn耗时比较长

**Comment by lyxbyr** [comment]
确实，我用nvprof --prinf-gpu-trace查看了DCNIm2colKernel 这个层加起来总共将近耗时40ms，对于速度进一步提升的话，请问下您有优化的建议和方向吗？

**Comment by hopef** [comment]
dcn的双线性插值执行的频率太高，耗时的主要问题，暂时没有更好的办法

**Comment by xjsxujingsong** [comment]
感觉可以参考centermask的backbone网络，只不过centermask要么是mask rcnn benchmark要么是detectron2，不好导出到onnx。还有c++处理多维矩阵比较麻烦，我觉得可以在python方便变换维度等等操作

**Comment by Qjizhi** [comment]
> 您好！thanks you amazing works
> 我想实现3d检测的推理，nuScenes_3Ddetction.fp32.trtmodel已经生成了，nuScenes_3Ddetction.onnx的输出只比你的nuScenes_3Dtracking.onnx少一个tracking的输出，输入是一张RGB图，我看了您的代码，里面的实现是基于2d检测和跟踪的,3d的检测或跟踪并没用实现，我现在取出了trt forward()后的7个输出了，
> eg:auto outHM = boundingbox3DDetect_->tensor("hm")等等;后面代码的三个for循环后不知道怎么改，因为先前您转的2d 检测和跟踪（onnx）最后输出都有个hm_pool, 所以里面的有句 if (*ohmptr == *ohmpoolptr && *ohmpoolptr > threshold)，我这个3d 检测onnx模型没有，我把这个if语句删了，打印这些std::cout << dx << " " << dy << " " << dw << " " << dh与图片的目标数不一致，请问下您怎么改，说的比较乱，如果直接用这些代码推理的话
> auto engine = TRTInfer::loadEngine("nuScenes_3Ddetction.fp32.trtmodel");
> float mean[3] = {0.485, 0.456, 0.406};
> float std[3] = {0.229, 0.224, 0.225};
> Mat image = imread("img.jpg");
> engine->input()->setNormMat(0, image, mean, std);
> engine->forward();
> engine->output(0)->print()后续的取值请问下怎么做？谢谢

如果你只需要做3d检测,可以参考我的Repo, 不过和这里作者的思路不太一样, 另外我现在在做 TensorRT 7 的3d centertrack 不知道有没有老哥搞过啊???
